custom_mutators/aflpp/aflpp.c

```diff
@@ -20,58 +56,97 @@ my_mutator_t *afl_custom_init(afl_state_t *afl, unsigned int seed) {
 
   }
 
-  if ((data->buf = malloc(MAX_FILE)) == NULL) {
+  data->fuzz_buf = (u8 *)malloc(MAX_FILE);
+  if (!data->fuzz_buf) {
 
-    perror("afl_custom_init alloc");
+    perror("afl_custom_init malloc");
     return NULL;
 
-  } else {
-
-    data->buf_size = MAX_FILE;
-
   }
-
   data->afl = afl;
 
   return data;
 
 }
 
-/* here we run the AFL++ mutator, which is the best! */
-
 size_t afl_custom_fuzz(my_mutator_t *data, uint8_t *buf, size_t buf_size,
-                       u8 **out_buf, uint8_t *add_buf, size_t add_buf_size,
+                       u8 **out_buf, uint8_t *add_buf,
+                       size_t add_buf_size,  // add_buf can be NULL
                        size_t max_size) {
 
-  if (max_size > data->buf_size) {
+  int size = (rand() % 100) + 1;
+  if (size > max_size) size = max_size;
+  
+  data->afl->from_llm = false;
+  data->afl->unique_id = -1;
 
-    u8 *ptr = realloc(data->buf, max_size);
+  /* the mutation, send request to LLM, then receive mutate seed */
+  message_seed_t my_msg;
+  int        msg = 200;
 
-    if (ptr) {
+  // Create or open the message queue
+  int msqid = msgget((key_t)1234, IPC_CREAT | 0666);
+  if (msqid == -1) {
+    printf("msgget() failed");
+  }
+  
+  // send the request (empty message)
+  memcpy(my_msg.data_buff, &msg, sizeof(int));
+  my_msg.data_type = TYPE_REQUEST;
+
+  int snd_status = msgsnd(msqid, &my_msg, 0, 0);
+  if (snd_status == -1) {
+    printf("request send failed");
+  }
+  // receive seed info from llm
+  clock_t start_time;
+  start_time = clock();
 
-      return 0;
+  // if run time exceed 0.1s then break and mutate default one
+  while (((double)(clock() - start_time)) / CLOCKS_PER_SEC < 0.05) {
+    int rcv_status = msgrcv(msqid, &my_msg, sizeof(message_seed_t) - sizeof(long), -2, 0);
 
+    if (rcv_status == -1 ) {
+      printf("RECEIVE ERROR %d \n",rcv_status);
+      break;
     } else {
-
-      data->buf = ptr;
-      data->buf_size = max_size;
-
+      // receive non-empty seed(uid+seed)
+      if (my_msg.data_type == TYPE_SEED){
+        printf("RECEIVED");
+        size_t hexLength = strlen(my_msg.data_buff);
+        size_t byteLength = hexLength / 2;
+        if (MAX_FILE < byteLength) {
+          printf("Buffer size %ld is too small for the hex string %ld.\n",MAX_FILE,byteLength);
+          break;
+        }
+
+        for (size_t i = 0, j = 0; i < hexLength; i += 2, j++) {
```

Comment:
Why don't you just send a buffer of bytes?

Reply:
It is send a buffer of bytes, and this part of codes is to write hex string to the seed buffer.